presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java

```diff
@@ -146,10 +146,35 @@ public SqlTask load(TaskId taskId)
                         },
                         maxBufferSize
                 );
+                newSqlTask.addSystemMemoryUsageListener(new UpdateSystemMemory(queryContexts.getUnchecked(taskId.getQueryId())));
+                return newSqlTask;
             }
         });
     }
 
+    private static final class UpdateSystemMemory
```

Comment:
move this class to `SqlTask` then you can remove the setter on `SqlTask` and just construct this in the constructor